### PR TITLE
feat(http): exclude caching for authenticated HTTP requests

### LINKED ISSUE #54745: provideClientHydration() shouldn't cache HttpClient GET requests with "Authorization" header
### Which @angular/* package(s) are relevant/related to the feature request?

common, platform-browser

### Description

After I activated `provideClientHydration()` in my web application, I have noticed that GET requests from HttpClient are being cached internally by Angular, even if they are **requests with "Authorization" headers**. It's taken me a whole day to figure it out.

I've then tested with `provideClientHydration(withNoHttpTransferCache())`, and the extrange behavior stops happening. Nonetheless, I get the benefit of HttpClient caching when hydration is activated for GET requests without "Authorization" header. That being said, I'm finally trying the following configuration, which I think should be by default:

```js
    provideClientHydration(withHttpTransferCacheOptions({
      filter: req => req.method === 'GET' && !req.headers.has('Authorization'),
    })),
```

### Proposed solution

Stop `provideClientHydration()`from caching HttpClient's GET requests with "Authorization" header by default.

### Alternatives considered

More documentation on how `provideClientHydration()` works along with HttpClient and caching.
Also, in the following links I don't get what the default behavior or filters are:
https://angular.io/api/platform-browser/withHttpTransferCacheOptions
https://angular.io/api/common/http/HttpTransferCacheOptions

### PR DESCRIPTION
**Pull request by alan-agius4**
This update modifies the transfer cache logic to prevent caching of HTTP requests that require authorization. To opt-out from this behaviour use the `includeRequestsWithAuthHeaders` option in `withHttpTransferCache`

BREAKING CHANGE: By default we now prevent caching of HTTP requests that require authorization . To opt-out from this behaviour use the `includeRequestsWithAuthHeaders` option in `withHttpTransferCache`.

Example:
```ts
withHttpTransferCache({
  includeRequestsWithAuthHeaders: true,
})
```

Closes: #54745

### CODE REVIEW

**Review by AndrewKushnir — COMMENTED** [review]
@alan-agius4 thanks for the PR 👍 I've added a comment about the name of the option.

We can also add a note about this in the SSR guide, WDYT? Here is a proposed change:

```
--- a/adev/src/content/guide/ssr.md
+++ b/adev/src/content/guide/ssr.md
@@ -71,7 +71,7 @@ Hydration is the process that restores the server side rendered application on t
 
 [`HttpClient`](api/common/http/HttpClient) cached outgoing network requests when running on the server. This information is serialized and transferred to the browser as part of the initial HTML sent from the server. In the browser, `HttpClient` checks whether it has data in the cache and if so, reuses it instead of making a new HTTP request during initial application rendering. `HttpClient` stops using the cache once an application becomes [stable](api/core/ApplicationRef#isStable) while running in a browser.
 
-`HttpClient` caches all `HEAD` and `GET` requests by default. You can configure the cache by using [`withHttpTransferCacheOptions`](/api/platform-browser/withHttpTransferCacheOptions) when providing hydration.
+By default, `HttpClient` caches all `HEAD` and `GET` requests which don't contain `Authorization` or `Proxy-Authorization` headers. You can override those settings by using [`withHttpTransferCacheOptions`](/api/platform-browser/withHttpTransferCacheOptions) when providing hydration.
 
 <docs-code language="typescript">
```

File: packages/common/http/src/transfer_cache.ts [hunk 1]
```diff
@@ -97,13 +100,13 @@ export function transferCacheInterceptorFn(
...
     (requestMethod === 'POST' && !globalOptions.includePostRequests && !requestOptions) ||
     (requestMethod !== 'POST' && !ALLOWED_METHODS.includes(requestMethod)) ||
-    // Do not cache request that require authorization
-    req.headers.has('authorization') ||
-    req.headers.has('proxy-authorization') ||
-    requestOptions === false ||
+    // Do not cache request that require authorization when includeAuthorizationRequests is falsey
+    (!globalOptions.includeAuthorizationRequests &&
+      (req.headers.has('authorization') || req.headers.has('proxy-authorization'))) ||
     globalOptions.filter?.(req) === false
```
Comment (JoostK) [inline packages/common/http/src/transfer_cache.ts:110]: Given that there's already global filtering capabilities available, do we really want to handle Authorization as special case?

Comment (alan-agius4) [reply packages/common/http/src/transfer_cache.ts:110]: The concept was to prioritize doing the "right" thing by default, which implies that `authorization` requests won't be cached by default.

Considering this, perhaps the global filter capabilities should execute before any other checks. If it returns `true`, subsequent internal checks could be bypassed.

Consequently, this approach could render the `includePostRequest` option unnecessary.

//cc @AndrewKushnir

File: packages/common/http/src/transfer_cache.ts [hunk 2]
```diff
@@ -100,10 +103,13 @@ export function transferCacheInterceptorFn(
...
   if (
     !isCacheActive ||
+    requestOptions === false ||
     // POST requests are allowed either globally or at request level
     (requestMethod === 'POST' && !globalOptions.includePostRequests && !requestOptions) ||
     (requestMethod !== 'POST' && !ALLOWED_METHODS.includes(requestMethod)) ||
-    requestOptions === false || //
+    // Do not cache request that require authorization when includeRequestsWithAuthHeaders is falsey
+    (!globalOptions.includeRequestsWithAuthHeaders &&
+      (req.headers.has('authorization') || req.headers.has('proxy-authorization'))) ||
```
Comment (AndrewKushnir) [inline packages/common/http/src/transfer_cache.ts:112]: nit: I'd propose moving the logic that checks for specific headers into a separate function:
```suggestion
    (!globalOptions.includeRequestsWithAuthHeaders && hasAuthHeaders(req)) ||
```

**Review by AndrewKushnir — APPROVED** [review]
Looks great, just 1 minor comment 👍

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 8eacb6e4b982a5aa23cfbf9078dc4e19d9466d73.

**Pull request closed without merging**